napari/layers/labels/labels.py

```diff
@@ -649,6 +666,10 @@ def _raw_to_displayed(self, raw):
         else:
             raise ValueError("Unsupported Color Mode")
 
+        if self.contour:
+            mask = np.logical_xor(raw, ndi.binary_erosion(raw))
+            image[np.invert(mask)] = 0
```

Comment:
The problem with the dense labels example is that if you take the raw image as bool, it becomes all 1s. Instead of `logical_xor` you probably want to compare the dilated and eroded images. See the [source code for `skimage.segmentation.find_boundaries`](https://github.com/scikit-image/scikit-image/blob/master/skimage/segmentation/boundaries.py#L48-L181) for inspiration.

More broadly, I'm worried that this code will become obsolete very quickly, since we are aiming to move the rendering code to vispy instead of doing it on the CPU...

Reply:
(*but*, I wouldn't be opposed to getting it in — it's a good vis mode to have. We'll just have to figure out how to do this in a shader. 😬)

Reply:
Yes I think fine to get in now and then move to shader, will be easier to have the rest of the implementation and then just port this one piece.

Reply:
> The problem with the dense labels example is that if you take the raw image as bool, it becomes all 1s. Instead of `logical_xor` you probably want to compare the dilated and eroded images. See the [source code for `skimage.segmentation.find_boundaries`](https://github.com/scikit-image/scikit-image/blob/master/skimage/segmentation/boundaries.py#L48-L181) for inspiration.

Thanks for the pointer, I'll have a look :+1:

Reply:
It seems to work as expected now, although it's a bit slow at creating the countours on the `labels-2d.py` example.

Reply:
@kir0ul but why are you looping over the labels? `find_boundaries` does it using one dilation, one erosion, and image arithmetic, rather than, in the current case, as many dilations as there are labels...

Reply:
I was basically not using the right functions for erosion and dilation (binary instead of grayscale), which is why I needed to loop over the labels, sorry image processing noob here :grimacing: 
Now it should be fixed and fast!